### PR TITLE
Sub Configs

### PR DESCRIPTION
**Pull request by ImToggle**
## Description
Allows developers to group multiple configs inside a single config (
An example is in TestConfig_Test.java

## Checklist
- [x] I made a clear description of what was changed
- [x] I stated why these changes were necessary
- [x] I updated documentation or said what needs to be updated
- [x] I made sure these changes are backwards compatible
- [x] This pull request is for one feature/bug fix

### CODE REVIEW

File: src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * <https://polyfrost.cc/legal/oneconfig/additional-terms>
+ */
+
+package cc.polyfrost.oneconfig.internal.config;
+
+import cc.polyfrost.oneconfig.config.Config;
+import cc.polyfrost.oneconfig.config.data.Mod;
+import cc.polyfrost.oneconfig.config.data.ModType;
+
+public class EmptyConfig extends Config {
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java]: why

File: src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java [hunk 2]
```diff
@@ -39,6 +39,7 @@
 
 public class ConfigCore {
     public static List<Mod> mods = new ArrayList<>();
+    public static List<Mod> subMods = new ArrayList<>();
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java]: use a hashmap

File: src/main/java/cc/polyfrost/oneconfig/gui/pages/ModsPage.java [hunk 3]
```diff
@@ -116,6 +117,18 @@ public void reloadMods() {
         for (Mod modData : ConfigCore.mods) {
             modCards.add(new ModCard(modData, modData.config == null || modData.config.enabled, false, OneConfigConfig.favoriteMods.contains(modData.name), this));
         }
+        if (this instanceof SubModsPage) {
+            Mod parent = ((SubModsPage) this).parentMod;
+            if (parent == null) return;
+            modCards.removeIf(modCard -> !parent.config.subMods.contains(modCard.getModData()) || parent.config.subModSettings == modCard.getModData());
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/gui/pages/ModsPage.java]: hashmap

File: src/main/java/cc/polyfrost/oneconfig/gui/elements/ModCard.java [hunk 4]
```diff
@@ -146,10 +148,30 @@ public void onClick() {
                     return;
                 }
             }
-            OneConfigGui.INSTANCE.openPage(new ModConfigPage(modData.defaultPage));
+
+            for (Mod mod : ConfigCore.subMods) {
+                if (modData == mod.config.subModSettings) {
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/gui/elements/ModCard.java]: hashmap

File: src/main/java/cc/polyfrost/oneconfig/config/Config.java [hunk 5]
```diff
@@ -270,6 +283,16 @@ protected final void generateOptionList(Object instance, Class<?> targetClass, O
                 optionNames.put(optionName, option);
             }
         }
+        if (!subMods.isEmpty()) {
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/config/Config.java:284]: this is a rather hacky way to get this to work. i mean this is v0 so i guess its fine..

**Review by nextdayy — CHANGES_REQUESTED** [review]

File: src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java (hunk 1 above)
Comment (ImToggle) [reply src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java]: so it doesnt save, and it wont be null

File: src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java (hunk 2 above)
Comment (ImToggle) [reply src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java]: Mod and what

Comment (nextdayy) [reply src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java]: mod to the sub mods

Comment (ImToggle) [reply src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java]: ``public static HashMap<Mod, List<Mod>> mods = new HashMap<>();``?

File: src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java (hunk 1 above)
Comment (nextdayy) [reply src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java]: think this is a pretty reasonable use-case for null

File: src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java (hunk 2 above)
Comment (nextdayy) [reply src/main/java/cc/polyfrost/oneconfig/internal/config/core/ConfigCore.java]: sure just dont store them in a centralized list like that. you are using it like a hashmap anyway

File: src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java (hunk 1 above)
Comment (ImToggle) [reply src/main/java/cc/polyfrost/oneconfig/internal/config/EmptyConfig.java]: have to add lots of null check if i set it to null

File: src/main/java/cc/polyfrost/oneconfig/config/Config.java [hunk 6]
```diff
@@ -260,6 +266,11 @@ protected final void generateOptionList(Object instance, Class<?> targetClass, O
                 else subcategory.bottomButtons.add(button);
             } else if (field.isAnnotationPresent(HUD.class)) {
                 HUDUtils.addHudOptions(page, field, instance, this);
+            } else if (field.isAnnotationPresent(SubConfig.class)) {
+                if (page != mod.defaultPage) continue;
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/config/Config.java:270]: ?

File: src/main/java/cc/polyfrost/oneconfig/config/data/Mod.java [hunk 7]
```diff
@@ -116,13 +118,13 @@ public int compareTo(@NotNull Mod mod) {
...
-    @Override
-    public boolean equals(Object obj) {
-        if (obj == null) return false;
-        if (obj == this) return true;
-        if (obj instanceof Mod) {
-            Mod mod = (Mod) obj;
-            return name.equals(mod.name);
-        } else return false;
-    }
+//    @Override
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/config/data/Mod.java:121]: ?

File: src/main/java/cc/polyfrost/oneconfig/gui/pages/Page.java [hunk 8]
```diff
@@ -45,7 +45,7 @@
  * A page is a 1056x728 rectangle of the GUI. It is the main content of the gui, and can be switched back and forwards easily. All the content of OneConfig is in a page.
  */
 public abstract class Page {
-    protected final String title;
+    protected String title;
```
Comment (nextdayy) [inline src/main/java/cc/polyfrost/oneconfig/gui/pages/Page.java:48]: does this have to be done?

**Review by nextdayy — COMMENTED** [review]
looks okay to me - dont love it but it is much better. as long as this is adequately tested and works then im okay to merge.

File: src/main/java/cc/polyfrost/oneconfig/config/Config.java (hunk 6 above)
Comment (ImToggle) [reply src/main/java/cc/polyfrost/oneconfig/config/Config.java:270]: get the SubConfig annotation

File: src/main/java/cc/polyfrost/oneconfig/config/data/Mod.java (hunk 7 above)
Comment (ImToggle) [reply src/main/java/cc/polyfrost/oneconfig/config/data/Mod.java:121]: just check if the object is equal

File: src/main/java/cc/polyfrost/oneconfig/gui/pages/Page.java (hunk 8 above)
Comment (ImToggle) [reply src/main/java/cc/polyfrost/oneconfig/gui/pages/Page.java:48]: to set the title, yes

**Review by Wyvest — APPROVED** [review]

**Pull request merged**